polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/EntityCacheGrantManager.java

```diff
@@ -0,0 +1,236 @@
+    this.delegateGrantManager = delegateGrantManager;
+    this.entityCache = entityCache;
+  }
+
+  @Override
+  public @NotNull PrivilegeResult grantUsageOnRoleToGrantee(
+      @NotNull PolarisCallContext callCtx,
+      @Nullable PolarisEntityCore catalog,
+      @NotNull PolarisEntityCore role,
+      @NotNull PolarisEntityCore grantee) {
+    try {
+      return delegateGrantManager.grantUsageOnRoleToGrantee(callCtx, catalog, role, grantee);
+    } finally {
+      LOGGER.debug("Invalidating cache for role {} and grantee {}", role, grantee);
+      entityCache.removeCacheEntry(role);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull PrivilegeResult revokeUsageOnRoleFromGrantee(
+      @NotNull PolarisCallContext callCtx,
+      @Nullable PolarisEntityCore catalog,
+      @NotNull PolarisEntityCore role,
+      @NotNull PolarisEntityCore grantee) {
+    try {
+      return delegateGrantManager.revokeUsageOnRoleFromGrantee(callCtx, catalog, role, grantee);
+    } finally {
+      LOGGER.debug("Invalidating cache for role {} and grantee {}", role, grantee);
+      entityCache.removeCacheEntry(role);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull PrivilegeResult grantPrivilegeOnSecurableToRole(
+      @NotNull PolarisCallContext callCtx,
+      @NotNull PolarisEntityCore grantee,
+      @Nullable List<PolarisEntityCore> catalogPath,
+      @NotNull PolarisEntityCore securable,
+      @NotNull PolarisPrivilege privilege) {
+    try {
+      return delegateGrantManager.grantPrivilegeOnSecurableToRole(
+          callCtx, grantee, catalogPath, securable, privilege);
+    } finally {
+      LOGGER.debug("Invalidating cache for securable {} and grantee {}", securable, grantee);
+      entityCache.removeCacheEntry(securable);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull PrivilegeResult revokePrivilegeOnSecurableFromRole(
+      @NotNull PolarisCallContext callCtx,
+      @NotNull PolarisEntityCore grantee,
+      @Nullable List<PolarisEntityCore> catalogPath,
+      @NotNull PolarisEntityCore securable,
+      @NotNull PolarisPrivilege privilege) {
+    try {
+      return delegateGrantManager.revokePrivilegeOnSecurableFromRole(
+          callCtx, grantee, catalogPath, securable, privilege);
+    } finally {
+      LOGGER.debug("Invalidating cache for securable {} and grantee {}", securable, grantee);
+      entityCache.removeCacheEntry(securable);
+      entityCache.removeCacheEntry(grantee);
+    }
+  }
+
+  @Override
+  public @NotNull LoadGrantsResult loadGrantsOnSecurable(
+      @NotNull PolarisCallContext callCtx, long securableCatalogId, long securableId) {
+    EntityCacheLookupResult lookupResult =
+        entityCache.getOrLoadEntityById(callCtx, securableCatalogId, securableId);
+    if (lookupResult == null || lookupResult.getCacheEntry() == null) {
```

Comment:
I think this causes an "API skew". We never call `loadGrantsOnSecurable` on `delegateGrantManager`, therefore we assume that the `entityCache` is somehow connected to the same source of data as the `delegateGrantManager` but their APIs are totally different.

Essentially this means that `PolarisMetaStoreManagerImpl.loadCachedEntryById` and `PolarisMetaStoreManagerImpl.loadGrantsOnSecurable()` must produce coherent results.

It does not look like we achieve any clean API separation, which is kind of implies by having the multitude of interfaces that `PolarisMetaStoreManager` extends.

Reply:
yeah, the API separation still needs a lot of work. Originally, I wanted to break the cache methods out of the `PolarisMetaStoreManager` entirely, but without the DI support, it was too hard to cast underlying impls to the right interfaces. 

But, in reality, this API skew has really always been there. It's just more clearly evident now that we're explicitly invoking `loadCachedEntryById` when `loadGrantsOnSecurable` is called. Previously, this was just hidden by the fact that the authorizer simply got passed the `ResolvedPolarisEntity` and the Resolver just loaded everything from the cache without any pretense of talking to the `PolarisMetaStoreManager`. 

Eventually, I think the `EntityCache` and the `PolarisRemoteCache` types need to be addressed, but first is to encapsulate their usage into more generalized types so that they can be changed without impacting the rest of the calling code.

Reply:
I'd like to leave this concern to the proposal in https://docs.google.com/document/d/1MNCdW-uKVZaR5Ua91FDwxhreBeGjPOoV-CyF0WOWwIg/edit?tab=t.0#heading=h.e7gcs110cqli